crates/zizmor/src/yaml_patch/mod.rs

```diff
@@ -474,6 +527,148 @@ fn indent_multiline_yaml(content: &str, base_indent: &str) -> String {
     result
 }
 
+fn handle_block_mapping_addition(
+    doc: &yamlpath::Document,
+    patch: &Patch,
+    content: &str,
+    feature: &yamlpath::Feature,
+    key: &str,
+    value: &serde_yaml::Value,
+) -> Result<String, Error> {
+    // Convert the new value to YAML string for block style handling
+    let new_value_str = if matches!(value, serde_yaml::Value::Sequence(_)) {
+        // For sequences, use flow-aware serialization to maintain consistency
+        serialize_yaml_value_for_flow_context(value)?
+    } else {
+        serialize_yaml_value(value)?
+    };
+    let new_value_str = new_value_str.trim_end(); // Remove trailing newline
+
+    // Check if we're adding to a list item by examining the path
+    let is_list_item = matches!(patch.route.last(), Some(RouteComponent::Index(_)));
+
+    // Determine the appropriate indentation
+    let indent = if is_list_item {
+        // For list items, we need to match the indentation of other keys in the same item
+        // The feature extraction gives us the content without the leading "- " part,
+        // so we need to use the leading whitespace of the step itself plus 2 spaces
+        let leading_whitespace = extract_leading_whitespace(content, feature.location.byte_span.0);
+        format!("{}  ", leading_whitespace)
+    } else {
+        // For regular mappings, add 2 spaces to current indentation
+        let leading_whitespace = extract_leading_whitespace(content, feature.location.byte_span.0);
+        format!("{}", leading_whitespace)
+    };
+
+    // Format the new entry
+    let final_entry = if let serde_yaml::Value::Mapping(mapping) = &value {
+        if mapping.is_empty() {
+            // For empty mappings, format inline
+            format!("\n{}{}: {}", indent, key, new_value_str)
+        } else {
+            // For non-empty mappings, format as a nested structure
+            let value_lines: Vec<&str> = new_value_str.lines().collect();
+            let mut result = format!("\n{}{}:", indent, key);
+            for line in value_lines.iter() {
+                if !line.trim().is_empty() {
+                    result.push('\n');
+                    result.push_str(&indent);
+                    result.push_str("  "); // 2 spaces for nested content
+                    result.push_str(line.trim_start());
+                }
+            }
+            result
+        }
+    } else if new_value_str.contains('\n') {
+        // Handle multiline values
+        let indented_value = indent_multiline_yaml(new_value_str, &indent);
+        format!("\n{}{}: {}", indent, key, indented_value)
+    } else {
+        format!("\n{}{}: {}", indent, key, new_value_str)
+    };
+
+    // Find the insertion point
+    let insertion_point = if is_list_item {
+        // For list items, we need to find the end of the actual step content,
+        // not including trailing comments that yamlpath may have included
+        find_step_content_end(content, &feature, &doc)
```

Comment:
Flagging for myself: I want to push this fix up into `yamlpath`, since it's arguably a deficiency in the feature abstraction that shouldn't need to be handled at the patching layer.

(More generally, it's arguably a bug in the tree-sitter YAML parser, which attaches comments in pretty unintuitive places. But the buck stops with `yamlpath` 🙂)

Reply:
I got halfway with this: pushing it all the way into `yamlpath` would be nontrivial, but `yamlpath::Document` now caries a `LineIndex` that makes this and other operations much simpler.